pkg/io/v2/providers/providers.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/fileblob"
+	"gocloud.dev/blob/gcsblob"
+	"gocloud.dev/blob/s3blob"
+	"gocloud.dev/gcp"
+	"golang.org/x/oauth2/google"
+)
+
+const (
+	// Required as long as paths without prefix are used
+	defaultStorageProviderName = "gs"
+
+	storageSeparator = "://"
+
+	providerFile = "file"
+	providerGS   = "gs"
+	providerS3   = "s3"
+)
+
+var storageProviders = []string{providerFile, providerGS, providerS3}
+
+func GetBucket(ctx context.Context, credentials []byte, path string) (*blob.Bucket, error) {
```

Comment:
Re
```
func GetBucket(ctx context.Context, credentials []byte, path string) (*blob.Bucket, error) {
	storageProvider, bucket, _, err := ParseStoragePath(path)
	if err != nil {
		return nil, err
	}
	if storageProvider == providerFile || len(credentials) == 0 {
		bkt, err := blob.OpenBucket(context.Background(), fmt.Sprintf("%s://%s", storageProvider, bucket))
		if err != nil {
			return nil, fmt.Errorf("error opening file bucket: %v", err)
		}
		return bkt, nil
	}
	
	switch storageProvider {
	case providerGS:
		return getGCSBucket(ctx, credentials, bucket)
	case providerS3:
		return getS3Bucket(ctx, credentials, bucket)
	default:
		return nil, fmt.Errorf("unknown storageProvider: %s", storageProvider)
	}
}
```
> If no credentials are given, gocloud can try to auto-discover based on e.g. gs:// google credentials, etc.. Every provider supported by gocloud would be supported as long as we blank-import it and as long as the credentials are given via one of the default ways of the used provider.

> We then only have to write code for providers which should be supported via our blog-store-credentials-file flag.

Yup, that sounds perfect. Please also add godocs to `GetBucket` explaining this and its args

Reply:
Done